### ISSUE TITLE
Update Developer Guidelines

### ISSUE THREAD

**Issue by borjasotomayor** [uchicago-cs/chiventure#418]
The [developer guidelines](https://github.com/uchicago-cs/chiventure/blob/master/docs/developer.md) need to be updated for 2020. Some of the updates are pretty routine (like the new directory structure), but we need to decide on how we're going to update the branching strategy, which did not work that well last year (encouraging large PRs instead of small incremental PRs).

I believe at the end of last year's 220, we discussed switching to the following model:

- There are no per-component branches (like wdl/dev, cli/dev, etc. last year).
- "dev" takes on the role that "master" had last year: there is a high bar to merge into "dev", but teams now create branches straight out of "dev" (instead of spininng a branch out of their component-specific dev branch, which then needs to be merged into "master")
- "dev" is treated as a "pre-master" branch, where feature branches are integrated. Merging to "dev" requires a review by a senior developer.
- Merging from "dev" to "master" represents a "stable" release of chiventure. It is initiated by the product manager, and has no bearing on points, grades, etc.

In this model, there doesn't seem to be a way for having "non-master" PRs like we did last year, but my impression is that these were rarely used, and the goal is to instead have plenty of small PRs into "dev" (with approval from a senior developer), so those should still provide an opportunity to rake in points.

Also, to encourage more frequent PRs, do we want to require them to have two reviews (one student and one senior dev) instead of three (two students and one senior dev)? We can encourage them to request additional reviews when necessary (e.g., if you're writing code that affects other components), but only require two reviews for the code to be merged.

**Comment by mpcsh** [comment]
(dropping in on this by request from Borja)

the branching model looks great here, this is definitely what we (+ Lydia) were discussing at the end of last year. the non-master PR thing I think was a miss, and definitely encouraged huge master PRs.

how would you feel about making the requirement be one senior dev approval plus approvals from at least one representative from each code-owning team? so, if you're writing code in your team's component, you need a teammate and a senior dev. if you're writing code that touches your team's component plus two other teams' components, you need a teammate, one person from each of the other two teams, and a senior dev. there's probably a better way to word that, but that'd be the gist.

I'd also explicitly call out that especially for cross-cutting code changes, a senior dev's approval does not negate teammate approvals. the senior devs will always have less context than teammates, and PRs shouldn't be a matter of getting them perfect _and then_ submitting to the senior dev - that should all happen in parallel.

**Comment by borjasotomayor** [comment]
I like the idea of requiring approvals from each code-owning team. It could also result in more points, since it represents a more complex PR.

**Comment by borjasotomayor** [comment]
Closed by #420

**Comment by mpcsh** [comment]
nice
